rust/worker/src/storage/s3.rs

```diff
@@ -155,56 +158,177 @@ impl S3Storage {
-                return Ok(());
+            .await
+            .map_err(|err| S3PutError::S3PutError(err.to_string()))?
+            .upload_id
+        {
+            Some(upload_id) => upload_id,
+            None => {
+                return Err(S3PutError::S3PutError(
+                    "Multipart upload creation response missing upload ID".to_string(),
+                ));
             }
-            Err(e) => match e {
-                SdkError::ServiceError(err) => {
-                    let inner_err = err.into_err();
-                    let err_string = format!(
-                        "S3 service error with code: {:?} and message: {:?}",
-                        inner_err.code(),
-                        inner_err.message()
-                    );
-                    tracing::error!("{}", err_string);
-                    return Err(S3PutError::S3PutError(err_string));
-                }
-                SdkError::DispatchFailure(e) => {
-                    tracing::error!("S3 Dispatch failure error {:?}", e);
-                    return Err(S3PutError::S3DispatchFailure);
-                }
-                _ => {
-                    tracing::error!("S3 Put Error: {}", e);
-                    return Err(S3PutError::S3PutError(e.to_string()));
-                }
-            },
+        };
+
+        let mut upload_parts = Vec::new();
+        for (part_number, offset, length) in
+            S3Storage::part_number_offset_length_iter(file_size, part_size_bytes)
+        {
+            let stream = ByteStream::read_from()
+                .path(path)
+                .offset(offset)
+                .length(Length::Exact(length))
+                .build()
+                .await
+                .map_err(|err| S3PutError::S3PutError(err.to_string()))?;
+
+            let upload_part_res = self
+                .client
+                .upload_part()
+                .key(key)
+                .bucket(&self.bucket)
+                .upload_id(&upload_id)
+                .body(stream)
+                .part_number(part_number)
+                .send()
+                .await
+                .map_err(|err| S3PutError::S3PutError(err.to_string()))?;
+
+            upload_parts.push(
+                CompletedPart::builder()
+                    .e_tag(upload_part_res.e_tag.unwrap_or_default())
+                    .part_number(part_number)
+                    .build(),
+            );
         }
+
+        self.client
+            .complete_multipart_upload()
+            .bucket(&self.bucket)
+            .key(key)
+            .multipart_upload(
+                CompletedMultipartUpload::builder()
+                    .set_parts(Some(upload_parts))
+                    .build(),
+            )
+            .upload_id(&upload_id)
+            .send()
+            .await
+            .map_err(|err| S3PutError::S3PutError(err.to_string()))?;
+
+        Ok(())
+    }
+
+    fn part_number_offset_length_iter(
```

Comment:
at first I was going to factor out more of the shared multipart upload logic and use an async closure parameter that returns a ByteStream given an offset and length
but was running into some lifetime/type issues so went with this simpler approach for now

Reply:
yeah that would have been nice

Reply:
figured out how to do this, much cleaner now and I know a little more about async rust :)

58a42fe2